### PR TITLE
Implemented keyword exclusion

### PR DESCRIPTION
**Pull request by softcookiepp**
Keywords can now be excluded from the grammar fields by passing a list of word strings to the wordBlacklist argument during grammar field initialization.

### CODE REVIEW

**Comment by HyperCrowd** [comment]
Hey, this is a great start

Mind if I use this chance to teach you some arcane JavaScript-isms?

**Comment by softcookiepp** [comment]
> Hey, this is a great start
> 
> Mind if I use this chance to teach you some arcane JavaScript-isms?

Sure, I'm all ears! And is there any missing functionality?

**Comment by HyperCrowd** [comment]
When can you join a Jitsi this week with the team to review?

**Comment by softcookiepp** [comment]
> When can you join a Jitsi this week with the team to review?

I'm usually available between 7:00pm-12:00am pacific standard time every day. Any day works for me as long as it is within that time interval.

File: grammarField.js [hunk 1]
```diff
@@ -20,13 +20,13 @@ const colorMappings = [
 class GrammarField {
 
   /**
-   *
+   * 
    */
-  constructor(sentences = [], times = []) {
+  constructor(sentences = [], times = [],wordBlacklist=[]) {
```
Comment (HyperCrowd) [inline grammarField.js]: Please add a space after the comma

And

Put spaces around the =

File: grammarField.js [hunk 2]
```diff
@@ -20,13 +20,13 @@ const colorMappings = [
...
     if (sentences.length !== times.length) {
       throw new RangeError('Words must be the same length as times');
     }
-
+    this.wordBlacklist=this.preprocessWordBlacklist(wordBlacklist);
```
Comment (HyperCrowd) [inline grammarField.js]: Please add spaces around =

File: grammarField.js [hunk 3]
```diff
@@ -35,55 +35,81 @@ class GrammarField {
...
-
+  
+  
+  
+  /**
+   * singularizes all the words in the blacklist and converts them to lowercase
+   */
+  preprocessWordBlacklist(bl=[])
+  {
+    var newBL=[];
```
Comment (HyperCrowd) [inline grammarField.js]: Please add spaces around the =

Change var to const

Also give variable names something readable and descript

File: grammarField.js [hunk 4]
```diff
@@ -35,55 +35,81 @@ class GrammarField {
...
+    
   }
-
+  
+  
+  
+  /**
+   * singularizes all the words in the blacklist and converts them to lowercase
+   */
+  preprocessWordBlacklist(bl=[])
```
Comment (HyperCrowd) [inline grammarField.js]: Please add spaces around the =

File: grammarField.js [hunk 5]
```diff
@@ -35,55 +35,81 @@ class GrammarField {
...
+  {
+    var newBL=[];
+    for(let i=0;i<bl.length;i++)
```
Comment (HyperCrowd) [inline grammarField.js]: Add a space after for

Add spaces between all operators (=, <, etc)

Add spaces after the ;

File: grammarField.js [hunk 6]
```diff
@@ -35,55 +35,81 @@ class GrammarField {
...
+    {
+      var word=bl[i];
```
Comment (HyperCrowd) [inline grammarField.js]: Change var to let

Please add spaces around the =

File: grammarField.js [hunk 7]
```diff
@@ -35,55 +35,81 @@ class GrammarField {
...
+      word=word.toLowerCase();
```
Comment (HyperCrowd) [inline grammarField.js]: Please add spaces around the =

File: grammarField.js [hunk 8]
```diff
@@ -35,55 +35,81 @@ class GrammarField {
...
+      word=pluralize.singular(word);
+      newBL.push(word);
```
Comment (HyperCrowd) [inline grammarField.js:55]: Idea
```
newBL.push(
 pluralize.singular(
   bl[i].toLowerCase()
 )
)
```

File: grammarField.js [hunk 9]
```diff
@@ -328,6 +354,10 @@ class GrammarField {
 
     // @TODO: create PNG
   }
+  
+  executeWordFilters()
```
Comment (HyperCrowd) [inline grammarField.js]: What does this function do?

File: grammarField.js [hunk 10]
```diff
@@ -35,55 +35,81 @@ class GrammarField {
...
         // Singularize all words to make context easier to group
-        const word = pluralize.singular(baseWord)
-
-        if (this.uniqueWords[word] === undefined) {
-          this.uniqueWords[word] = {
+        const word = pluralize.singular(baseWord);
+        
+        // condition for word filter
+        // everything relating to the word parameter calculation was put inside this condition block in order to avoid errors
+        if(this.wordBlacklist.includes(word) == false)
```
Comment (HyperCrowd) [inline grammarField.js]: Includes will check if the word is a regular expression first, so use this `indexOf` instead

```js
if (this.wordBlacklist.indexOf(word) === -1) {
```

**Comment by softcookiepp** [comment]
Got it, all the requested changes are finished. I also added a test script, wordBlacklistTest.js, to the tests folder so the implementation can be tested quickly. So it all still works after making changes.

**Review by HyperCrowd — APPROVED** [review]
Excellent.  You even added a test.  Very polite, thank you! :D

**Pull request merged**